atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCandidateRowsForSweeping.java

```diff
@@ -0,0 +1,114 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.sweep;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Lists;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweepingRequest;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.cassandra.CqlExecutor;
+
+public class GetCandidateRowsForSweeping {
+
+    private static final int DEFAULT_TIMESTAMPS_BATCH_SIZE = 10_000;
+    private static final int CONSERVATIVE_MAX_VALUES_BATCH_SIZE = 100;
+
+    private final ValuesLoader valuesLoader;
+    private final CqlExecutor cqlExecutor;
+    private final TableReference table;
+    private final CandidateCellForSweepingRequest request;
+    private final int timestampsBatchSize;
+    private final int valuesBatchSize;
+
+    private List<CellWithTimestamps> cellTimestamps;
+    private Set<Cell> cellsWithEmptyValues;
+
+    public GetCandidateRowsForSweeping(
+            ValuesLoader valuesLoader,
+            CqlExecutor cqlExecutor,
+            TableReference table,
+            CandidateCellForSweepingRequest request) {
+        this.table = table;
+        this.cqlExecutor = cqlExecutor;
+        this.request = request;
+        this.valuesLoader = valuesLoader;
+
+        this.timestampsBatchSize = request.batchSizeHint().orElse(DEFAULT_TIMESTAMPS_BATCH_SIZE);
+        // TODO(nziebart): this should probably be configurable
+        this.valuesBatchSize = Math.max(1, Math.min(CONSERVATIVE_MAX_VALUES_BATCH_SIZE, timestampsBatchSize / 32));
+    }
+
+    /**
+     * Fetches a batch of candidate rows. The returned {@link CandidateRowForSweeping}s are ordered by row.
+     */
+    public List<CandidateRowForSweeping> execute() {
+        fetchCellTimestamps();
+
+        findCellsWithEmptyValuesIfNeeded();
+
+        return convertToOrderedSweepCandidateRows();
+    }
+
+    private void fetchCellTimestamps() {
+        cellTimestamps = new GetCellTimestamps(cqlExecutor, table, request.startRowInclusive(), timestampsBatchSize)
+                .execute();
+    }
+
+    public void findCellsWithEmptyValuesIfNeeded() {
+        if (!request.shouldCheckIfLatestValueIsEmpty()) {
+            cellsWithEmptyValues = Collections.emptySet();
+            return;
+        }
+
+        cellsWithEmptyValues = new GetEmptyLatestValues(cellTimestamps, valuesLoader, table,
```

Comment:
This is technically incorrect. Since you load "cellsWithEmptyValues" in a separate request, the latest timestamps at that point might not match the latest timestamps you got during fetchCellTimestamps().

Reply:
Also this is an extra round trip to Cassandra. If you are loading values anyways, why not do it during the first call.

Reply:
Note that this was also the case before (in GetCandidateCellsForSweepingShim). My assumption was that it is not actually I incorrect; if the latest value is empty, it's ok to delete any previous value.

Reply:
The problem isn't that you cannot load just the latest value

Reply:
Yes this was the case before (and is a known problem), but why keep the bug if we can get rid of it.

> if the latest value is empty, it's ok to delete any previous value.
I don't think that's true. You need to first check if the deleted value is committed.